Add validations to prevent delete on read only models

kbrock opened the pull request:
We currently have some logic in the API for checking read_only records.

This PR moves that logic into the model using `supports` so we can make those api controllers more consistent.

1. We need to update the api controllers to check `supports`, but this is pretty standard across all our api controllers.
2. A number of models will respond `supports?(:delete) == false` because we have not added `supports` to every model. If we mix this module into `ApplicationRecord` we will not be able to leverage `respond_to?(:supports?)` to come up with generic logic. Don't think this is a problem but putting it out there.
3. Added an entry for all `read_only` models. A few of these models may be better served with standard rails validations.


|    table_name     |   column_name||
|-------------------|------------|---|
| advanced_settings | read_only|
| classifications   | read_only|
| conditions        | read_only|
| dialog_fields     | read_only|
| host_storages     | read_only|means something else|
| miq_alerts        | read_only|
| miq_policies      | read_only|
| miq_sets          | read_only|
| miq_user_roles    | read_only|
| miq_widgets       | read_only|

Fryguy commented:
Some of these are surprising, namely advanced_settings and hosts_storages - would like @agrare to weigh in on those.

agrare commented on app/models/host_storage.rb:
```diff
@@ -2,6 +2,9 @@ class HostStorage < ApplicationRecord
   belongs_to :host
   belongs_to :storage
 
+  supports(:update) { _("Can not edit read only model") if read_only? }
+  supports(:delete) { _("Can not delete read only model") if read_only? }
+
```
This means that the _storage_ is read only from the POV of the host, not that the _record_ is read-only.

kbrock replied:
thnx

Fryguy commented on app/models/advanced_setting.rb:
```diff
@@ -1,6 +1,9 @@
 class AdvancedSetting < ApplicationRecord
   belongs_to :resource, :polymorphic => true
 
+  supports(:update) { _("Can not edit read only model") if read_only? }
+  supports(:delete) { _("Can not delete read only model") if read_only? }
```
- Not sure on using the word "model" here, because that is a developer term which probably doesn't make sense to an end user.  What is the wording currently like in the UI?
- `Can not` -> `Cannot`

kbrock replied:
I was worried that this would add too many translations if they were all different. and didn't think the exception wording mattered.

I will fix

Fryguy commented on app/models/dialog_field.rb:
```diff
@@ -41,6 +41,9 @@ class DialogField < ApplicationRecord
   after_initialize :default_resource_action
 
+  supports(:update) { _("Can not edit read only model") if read_only? }
+  supports(:delete) { _("Can not delete read only model") if read_only? }
```
Strange that we have this for Dialog Field, but not other Dialog models?

kbrock replied:
So lets just skip on this one?

kbrock commented:
update:
- moved supports over to validators
- added custom matcher: `expect().to be_deleted` (may have modified too many files on this one)

update:
- removed host storages validator
- cop spacing fix
- added model names rather than "model"
- fix spec that expected model name rather than model.

unchanged:
- cop about cattr. we don't seem to use them anywhere, so I'm not sure if I should pay attention to this.
- Do we want to change the term of read only to "default" or "seed" or something like that? No reason to add another term.
- do we want to drop validation on dialog fields?

kbrock commented:
update:
- fixed spacing
- spec for seeding
- Used `EvmDatabase` instead of `described_class` because it was too difficult for me to read with the generic clause

kbrock commented:
@miq-bot cross-repo-tests /all

agrare commented:
@kbrock wonder if a mixin for the `reject_if_read_only` could help reduce the duplication?

kbrock commented:
update:
- moved `reject_if_readonly` into a mixin

agrare commented on app/models/mixins/read_only_mixin.rb:
```diff
@@ -0,0 +1,17 @@
+  extend ActiveSupport::Concern
+
+  included do
+    before_destroy    :reject_if_read_only
```
Minor but some extra padding here (assume was copied from a model where this was lined up right)

kbrock commented:
update:
- fix cops with whitespace

Fryguy commented on app/models/mixins/read_only_mixin.rb:
```diff
@@ -0,0 +1,17 @@
+  def reject_if_read_only
+    if read_only? && !EvmDatabase.seeding?
+      model_name = self.class.name.pluralize.underscore.tr('_', ' ').gsub('miq ', '')
+      errors.add(:base, _("Read only #{model_name} cannot be deleted."))
```
This is not a valid i18n structure...should be:

```suggestion
      errors.add(:base, _("Read only %{model_name} cannot be deleted.") % model_name)
```

Fryguy replied:
Also I'm not sure if that reads correctly?  The model name itself should probably _also_ be translated.  I think you have to use the Dictionary for that.

Fryguy commented on lib/evm_database.rb:
```diff
@@ -58,15 +58,29 @@ def self.seedable_plugin_classes
     Vmdb::Plugins.flat_map { |p| p.try(:seedable_classes) }.compact
   end
 
+  @@seeding = false
+  def self.seeding?
+    defined?(@@seeding) && @@seeding
+  end
+
+  def self.with_seed
+    @@seeding = true
+    yield
+  ensure
+    @@seeding = false
+  end
```
This should probably use class instance vars as per rubocop.  Or perhaps `class_attribute`
